File: runtime/src/main/java/com/siimkinks/sqlitemagic/CompiledSelect1Impl.java
package com.siimkinks.sqlitemagic;

import android.database.Cursor;

import com.siimkinks.sqlitemagic.Query.DatabaseQuery;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.sqlite.db.SupportSQLiteDatabase;
import androidx.sqlite.db.SupportSQLiteStatement;

import static com.siimkinks.sqlitemagic.CompiledSelectImpl.CompiledFirstSelectImpl.addTakeFirstLimitClauseIfNeeded;
import static com.siimkinks.sqlitemagic.CompiledSelectImpl.createQueryObservable;
import static com.siimkinks.sqlitemagic.SqlUtil.bindAllArgsAsStrings;
import static java.lang.System.nanoTime;
import static java.util.concurrent.TimeUnit.NANOSECONDS;

final class CompiledSelect1Impl<T, S> extends DatabaseQuery<List<T>, T> implements CompiledSelect<T, S> {
  @NonNull
  final String sql;
  @Nullable
  final String[] args;
  @NonNull
  final Column<?, T, ?, ?, ?> selectedColumn;
  @NonNull
  final String[] observedTables;

  CompiledSelect1Impl(@NonNull String sql,
                      @Nullable String[] args,
                      @NonNull DbConnectionImpl dbConnection,
                      @NonNull final Column<?, T, ?, ?, ?> selectedColumn,
                      @NonNull String[] observedTables) {
    super(dbConnection, new Mapper<T>() {
      @Override
      public T apply(@NonNull Cursor cursor) {
        return selectedColumn.getFromCursor(cursor);
      }
    });
    this.sql = sql;
    this.args = args;
    this.selectedColumn = selectedColumn;
    this.observedTables = observedTables;
  }

  @NonNull
  @Override
  Cursor rawQuery(boolean inStream) {
    super.rawQuery(inStream);
    final SupportSQLiteDatabase db = dbConnection.getReadableDatabase();
    final long startNanos = nanoTime();
    final Cursor cursor = db.query(sql, args);
    if (SqliteMagic.LOGGING_ENABLED) {
      final long queryTimeInMillis = NANOSECONDS.toMillis(nanoTime() - startNanos);
      LogUtil.logQueryTime(queryTimeInMillis, observedTables, sql, args);
    }
    return FastCursor.tryCreate(cursor);
  }

  @Override
  List<T> map(@NonNull Cursor cursor) {
    try {
      final int rowCount = cursor.getCount();
      if (rowCount == 0) {
        return Collections.emptyList();
      }
      final Column<?, T, ?, ?, ?> selectedColumn = this.selectedColumn;
      final ArrayList<T> values = new ArrayList<>(rowCount);
      while (cursor.moveToNext()) {
        values.add(selectedColumn.<T>getFromCursor(cursor));
      }
      return values;
    } finally {
      cursor.close();
    }
  }

  @Override
  public String toString() {
    return "[Select1<List>; sql=" + sql + "]";
  }

  @NonNull
  @Override
  public List<T> execute() {
    return map(rawQuery(false));
  }

  @NonNull
  @Override
  public ListQueryObservable<T> observe() {
    return new ListQueryObservable<>(createQueryObservable(observedTables, this));
  }

  @NonNull
  @Override
  public CompiledFirstSelect<T, S> takeFirst() {
    return new CompiledFirstSelect1Impl<>(this, dbConnection);
  }

  @NonNull
  @Override
  public CompiledCountSelect<S> count() {
    return new CompiledSelectImpl.CompiledCountSelectImpl<>(sql, args, dbConnection, observedTables);
  }

  @NonNull
  @Override
  public CompiledCursorSelect<T, S> toCursor() {
    return new CompiledCursorSelect1Impl<>(this, dbConnection);
  }

  static final class CompiledFirstSelect1Impl<T, S> extends DatabaseQuery<T, T> implements CompiledFirstSelect<T, S> {
    @NonNull
    private final SupportSQLiteStatement selectStm;
    @NonNull
    final String sql;
    @Nullable
    final String[] args;
    @NonNull
    final Column<?, T, ?, ?, ?> selectedColumn;
    @NonNull
    final String[] observedTables;

    CompiledFirstSelect1Impl(@NonNull CompiledSelect1Impl<T, S> compiledSelect,
                             @NonNull DbConnectionImpl dbConnection) {
      super(dbConnection, null);
      final String sql = addTakeFirstLimitClauseIfNeeded(compiledSelect.sql);
      final String[] args = compiledSelect.args;
      final SupportSQLiteStatement selectStm = dbConnection.compileStatement(sql);
      bindAllArgsAsStrings(selectStm, args);
      this.selectStm = selectStm;
      this.sql = sql;
      this.args = args;
      this.selectedColumn = compiledSelect.selectedColumn;
      this.observedTables = compiledSelect.observedTables;
    }

    @Override
    T map(@Nullable Cursor __) {
      return execute();
    }

    @Nullable
    @Override
    public T execute() {
      final T val;
      final long startNanos;
      synchronized (selectStm) {
        startNanos = nanoTime();
        val = selectedColumn.getFromStatement(selectStm);
      }
      if (SqliteMagic.LOGGING_ENABLED) {
        final long queryTimeInMillis = NANOSECONDS.toMillis(nanoTime() - startNanos);
        LogUtil.logQueryTime(queryTimeInMillis, observedTables, sql, args);
      }
      return val;
    }

    @NonNull
    @Override
    public SingleItemQueryObservable<T> observe() {
      return new SingleItemQueryObservable<>(createQueryObservable(observedTables, this));
    }

    @Override
    public String toString() {
      return "[TAKE FIRST 1;sql=" + sql + "]";
    }
  }

  static final class CompiledCursorSelect1Impl<T, S> extends DatabaseQuery<Cursor, T> implements CompiledCursorSelect<T, S> {
    @NonNull
    private final String sql;
    @Nullable
    private final String[] args;
    @NonNull
    final Column<?, T, ?, ?, ?> selectedColumn;
    @NonNull
    private final String[] observedTables;

    CompiledCursorSelect1Impl(@NonNull CompiledSelect1Impl<T, S> compiledSelect,
                              @NonNull DbConnectionImpl dbConnection) {
      super(dbConnection, null);
      this.sql = compiledSelect.sql;
      this.args = compiledSelect.args;
      this.selectedColumn = compiledSelect.selectedColumn;
      this.observedTables = compiledSelect.observedTables;
    }

    @Nullable
    @Override
    public T getFromCurrentPosition(@NonNull Cursor cursor) {
      if (cursor instanceof FastCursor) {
        final FastCursor fastCursor = (FastCursor) cursor;
        fastCursor.syncWith(cursor);
        return selectedColumn.getFromCursor(fastCursor);
      }
      return selectedColumn.getFromCursor(cursor);
    }

    @NonNull
    @Override
    Cursor rawQuery(boolean inStream) {
      super.rawQuery(inStream);
      final SupportSQLiteDatabase db = dbConnection.getReadableDatabase();
      final long startNanos = nanoTime();
      final Cursor cursor = db.query(sql, args);
      if (SqliteMagic.LOGGING_ENABLED) {
        final long queryTimeInMillis = NANOSECONDS.toMillis(nanoTime() - startNanos);
        LogUtil.logQueryTime(queryTimeInMillis, observedTables, sql, args);
      }
      return FastCursor.tryCreate(cursor);
    }

    @Override
    Cursor map(@NonNull Cursor cursor) {
      return cursor;
    }

    @NonNull
    @Override
    public Cursor execute() {
      return rawQuery(false);
    }

    @NonNull
    @Override
    public SingleItemQueryObservable<Cursor> observe() {
      return new SingleItemQueryObservable<>(createQueryObservable(observedTables, this));
    }

    @Override
    public String toString() {
      return "[CURSOR 1; sql=" + sql + "]";
    }
  }
}

File: runtime/src/main/kotlin/com/siimkinks/sqlitemagic/CompiledSelect1Impl.kt
package com.siimkinks.sqlitemagic

import android.database.Cursor
import com.siimkinks.sqlitemagic.CompiledSelectImpl.Companion.createQueryObservable
import com.siimkinks.sqlitemagic.CompiledSelectImpl.CompiledCountSelectImpl
import com.siimkinks.sqlitemagic.CompiledSelectImpl.CompiledFirstSelectImpl.Companion.addTakeFirstLimitClauseIfNeeded
import com.siimkinks.sqlitemagic.Query.DatabaseQuery
import java.util.concurrent.TimeUnit.NANOSECONDS

@Suppress("UNCHECKED_CAST")
internal class CompiledSelect1Impl<T, S>(
  @JvmField
  val sql: String,
  @JvmField
  val args: Array<String>?,
  dbConnection: DbConnectionImpl?,
  @JvmField
  val selectedColumn: Column<*, T, *, *, *>,
  @JvmField
  val observedTables: Array<String>
) : DatabaseQuery<List<T>, T>(
  dbConnection = dbConnection,
  mapper = Mapper { cursor ->
    selectedColumn.getFromCursor<T>(cursor) as T
  }
), CompiledSelect<T, S> {
  override fun rawQuery(
    inStream: Boolean,
    dbConnection: DbConnectionImpl
  ): Cursor {
    super.rawQuery(
      inStream = inStream,
      dbConnection = dbConnection
    )
    val db = dbConnection.readableDatabase
    val startNanos = System.nanoTime()
    val cursor = SqlUtil.query(db, sql, args)
    if (SqliteMagic.LOGGING_ENABLED) {
      val queryTimeInMillis = NANOSECONDS.toMillis(System.nanoTime() - startNanos)
      LogUtil.logQueryTime(queryTimeInMillis, observedTables, sql, args)
    }
    return FastCursor.tryCreate(cursor)
  }

  override fun map(
    cursor: Cursor?,
    dbConnection: DbConnectionImpl
  ): List<T> {
    checkNotNull(cursor).use { cursor ->
      val rowCount = cursor.count
      if (rowCount == 0) {
        return emptyList()
      }
      val selectedColumn = selectedColumn
      return buildList(rowCount) {
        while (cursor.moveToNext()) {
          add(selectedColumn.getFromCursor<T>(cursor) as T)
        }
      }
    }
  }

  override fun toString() = "[Select1<List>; sql=$sql]"

  override fun execute() = resolveConnection()
    .let { dbConnection ->
      map(
        cursor = rawQuery(
          inStream = false,
          dbConnection = dbConnection
        ),
        dbConnection = dbConnection
      )
    }

  override fun observe() = ListQueryObservable(
    createQueryObservable(
      observedTables = observedTables,
      query = this
    )
  )

  override fun takeFirst() = CompiledFirstSelect1Impl(
    compiledSelect = this,
    dbConnection = dbConnection
  )

  override fun count(): CompiledCountSelect<S> = CompiledCountSelectImpl(
    parentSql = sql,
    args = args,
    dbConnection = dbConnection,
    observedTables = observedTables
  )

  override fun toCursor() = CompiledCursorSelect1Impl(
    compiledSelect = this,
    dbConnection = dbConnection
  )

  internal class CompiledFirstSelect1Impl<T, S>(
    compiledSelect: CompiledSelect1Impl<T, S>,
    dbConnection: DbConnectionImpl?
  ) : DatabaseQuery<T, T>(
    dbConnection = dbConnection,
    mapper = null
  ), CompiledFirstSelect<T, S> {
    @JvmField
    val sql = addTakeFirstLimitClauseIfNeeded(compiledSelect.sql)

    @JvmField
    val args = compiledSelect.args

    @JvmField
    val selectedColumn = compiledSelect.selectedColumn

    @JvmField
    val observedTables = compiledSelect.observedTables

    private val selectStatement = when {
      compiledSelect.selectedColumn.valueParser.supportsStatementParsing() -> ConnectionStatement(
        sql = sql,
        args = args,
        initialConnection = dbConnection
      )
      else -> null
    }

    override fun map(
      cursor: Cursor?,
      dbConnection: DbConnectionImpl
    ) = execute(dbConnection)

    override fun execute(): T? = execute(resolveConnection())

    private fun execute(dbConnection: DbConnectionImpl): T? =
      when (val statement = selectStatement) {
        null -> executeFromCursor(dbConnection)
        else -> executeFromStatement(
          selectStatement = statement,
          dbConnection = dbConnection
        )
      }

    private fun executeFromStatement(
      selectStatement: ConnectionStatement,
      dbConnection: DbConnectionImpl
    ): T? = selectStatement.execute(
      dbConnection = dbConnection,
      operation = { statement ->
        val startNanos = System.nanoTime()
        val result = selectedColumn.getFromStatement<T>(statement)
        if (SqliteMagic.LOGGING_ENABLED) {
          val queryTimeInMillis = NANOSECONDS.toMillis(System.nanoTime() - startNanos)
          LogUtil.logQueryTime(queryTimeInMillis, observedTables, sql, args)
        }
        result
      }
    )

    private fun executeFromCursor(dbConnection: DbConnectionImpl): T? {
      val startNanos = System.nanoTime()
      val cursor = FastCursor.tryCreate(
        SqlUtil.query(dbConnection.readableDatabase, sql, args)
      )
      val result = cursor.use {
        when {
          it.moveToNext() -> selectedColumn.getFromCursor<T>(it)
          else -> null
        }
      }
      if (SqliteMagic.LOGGING_ENABLED) {
        val queryTimeInMillis = NANOSECONDS.toMillis(System.nanoTime() - startNanos)
        LogUtil.logQueryTime(queryTimeInMillis, observedTables, sql, args)
      }
      return result
    }

    override fun observe() = SingleItemQueryObservable(
      createQueryObservable(
        observedTables = observedTables,
        query = this
      )
    )

    override fun toString() = "[TAKE FIRST 1;sql=$sql]"
  }

  internal class CompiledCursorSelect1Impl<T, S>(
    compiledSelect: CompiledSelect1Impl<T, S>,
    dbConnection: DbConnectionImpl?
  ) : DatabaseQuery<Cursor, T>(
    dbConnection = dbConnection,
    mapper = null
  ), CompiledCursorSelect<T, S> {
    @JvmField
    val selectedColumn = compiledSelect.selectedColumn
    private val sql = compiledSelect.sql
    private val args = compiledSelect.args
    private val observedTables = compiledSelect.observedTables

    override fun getFromCurrentPosition(cursor: Cursor): T? {
      if (cursor is FastCursor) {
        cursor.syncWith(cursor)
        return selectedColumn.getFromCursor(cursor)
      }
      return selectedColumn.getFromCursor(cursor)
    }

    override fun rawQuery(
      inStream: Boolean,
      dbConnection: DbConnectionImpl
    ): Cursor {
      super.rawQuery(
        inStream = inStream,
        dbConnection = dbConnection
      )
      val db = dbConnection.readableDatabase
      val startNanos = System.nanoTime()
      val cursor = SqlUtil.query(db, sql, args)
      if (SqliteMagic.LOGGING_ENABLED) {
        val queryTimeInMillis = NANOSECONDS.toMillis(System.nanoTime() - startNanos)
        LogUtil.logQueryTime(queryTimeInMillis, observedTables, sql, args)
      }
      return FastCursor.tryCreate(cursor)
    }

    override fun map(
      cursor: Cursor?,
      dbConnection: DbConnectionImpl
    ) = cursor

    override fun execute() = checkNotNull(rawQuery(false))

    override fun observe() = SingleItemQueryObservable(
      createQueryObservable(
        observedTables = observedTables,
        query = this
      )
    )

    override fun toString() = "[CURSOR 1; sql=$sql]"
  }
}
